File: software/mobile/android/core-model/src/main/kotlin/com/avlyubimov/svc/core/model/RideDashboard.kt
package com.avlyubimov.svc.core.model

enum class RideDataState {
    VALID,
    STALE,
    DEGRADED,
    INVALID,
    UNAVAILABLE,
}

data class RideValue<T>(
    val value: T?,
    val unit: String,
    val state: RideDataState,
) {
    val displayValue: T?
        get() = when (state) {
            RideDataState.VALID,
            RideDataState.DEGRADED,
            -> value
            RideDataState.STALE,
            RideDataState.INVALID,
            RideDataState.UNAVAILABLE,
            -> null
        }
}

fun <T> Measurement<T>.toRideValue(): RideValue<T> {
    val state = when {
        stale -> RideDataState.STALE
        !valid || value == null -> if (quality == MeasurementQuality.INVALID) {
            RideDataState.INVALID
        } else {
            RideDataState.UNAVAILABLE
        }
        quality == MeasurementQuality.GOOD -> RideDataState.VALID
        quality == MeasurementQuality.DEGRADED -> RideDataState.DEGRADED
        quality == MeasurementQuality.INVALID -> RideDataState.INVALID
        else -> RideDataState.UNAVAILABLE
    }
    return RideValue(value = value, unit = unit, state = state)
}

enum class RideGear(val displayValue: String) {
    NEUTRAL("N"),
    FIRST("1"),
    SECOND("2"),
    THIRD("3"),
    FOURTH("4"),
    FIFTH("5"),
    SIXTH("6"),
    BETWEEN("BETWEEN"),
    UNAVAILABLE("—"),
}

enum class TachometerZone {
    NORMAL,
    WARNING,
    RED,
    UNAVAILABLE,
}

object TachometerZoneResolver {
    fun zone(rpm: Double?, profile: VehiclePerformanceProfile): TachometerZone {
        if (rpm == null || profile.tachometerScaleMaxRpm == null) {
            return TachometerZone.UNAVAILABLE
        }
        if (profile.redZoneStartRpm?.let { rpm >= it } == true) {
            return TachometerZone.RED
        }
        if (profile.warningStartRpm?.let { rpm >= it } == true) {
            return TachometerZone.WARNING
        }
        return TachometerZone.NORMAL
    }

    fun zoneWithHysteresis(
        rpm: Double?,
        profile: VehiclePerformanceProfile,
        previous: TachometerZone,
        hysteresisRpm: Double = 80.0,
    ): TachometerZone {
        if (rpm == null) return TachometerZone.UNAVAILABLE
        val raw = zone(rpm, profile)
        return when (previous) {
            TachometerZone.RED -> {
                if (
                    profile.redZoneStartRpm?.let {
                        rpm >= it - hysteresisRpm
                    } == true
                ) {
                    TachometerZone.RED
                } else {
                    raw
                }
            }
            TachometerZone.WARNING -> when {
                profile.redZoneStartRpm?.let { rpm >= it } == true ->
                    TachometerZone.RED
                profile.warningStartRpm?.let {
                    rpm >= it - hysteresisRpm
                } == true ->
                    TachometerZone.WARNING
                else -> raw
            }
            TachometerZone.NORMAL,
            TachometerZone.UNAVAILABLE,
            -> raw
        }
    }
}

data class LeanExtrema(
    val maximumLeftDegrees: Double = 0.0,
    val maximumRightDegrees: Double = 0.0,
) {
    fun observe(leanDegrees: Double?): LeanExtrema = when {
        leanDegrees == null -> this
        leanDegrees < 0 -> copy(
            maximumLeftDegrees = maxOf(maximumLeftDegrees, kotlin.math.abs(leanDegrees)),
        )
        else -> copy(maximumRightDegrees = maxOf(maximumRightDegrees, leanDegrees))
    }

    fun resetIfStationary(speed: RideValue<Double>): LeanExtrema =
        if (speed.displayValue == 0.0) LeanExtrema() else this
}

data class MotorcycleMountingTransform(
    val rollDegrees: Double,
    val pitchDegrees: Double,
    val yawDegrees: Double,
    val zeroOffsetDegrees: Double,
)

object MotorcycleCalibrationPolicy {
    fun canCalibrate(speed: RideValue<Double>, demoMode: Boolean): Boolean =
        demoMode && speed.displayValue == 0.0
}

enum class RideAlertSeverity(val priority: Int) {
    INFO(0),
    WARNING(1),
    CRITICAL(2),
}

data class RideAlert(
    val id: String,
    val title: String,
    val severity: RideAlertSeverity,
)

data class RideDashboardState(
    val speed: RideValue<Double>,
    val engineRpm: RideValue<Double>,
    val gear: RideGear,
    val leanAngle: RideValue<Double>,
    val engineTemperature: RideValue<Double>,
    val fuelLevel: RideValue<Double>,
    val ambientTemperature: RideValue<Double>,
    val ambientLight: RideValue<Double>,
    val batteryVoltage: RideValue<Double>,
    val totalCurrent: RideValue<Double>,
    val canState: RideValue<String>,
    val bleLabel: String,
    val bleState: RideDataState,
    val tachometerZone: TachometerZone,
    val alerts: List<RideAlert>,
    val profile: VehiclePerformanceProfile,
) {
    val activeAlert: RideAlert?
        get() = alerts.maxWithOrNull(
            compareBy<RideAlert> { it.severity.priority }.thenByDescending { it.id },
        )

    companion object {
        fun build(
            telemetry: TelemetrySnapshot,
            connectionState: ConnectionState,
            isConnecting: Boolean,
            profile: VehiclePerformanceProfile,
            leanCalibrated: Boolean = false,
        ): RideDashboardState {
            val speed = telemetry.vehicle.speed.toRideValue()
            val rpm = telemetry.vehicle.engineRpm.toRideValue()
            val rawLean = telemetry.leanAngle.toRideValue()
            val lean = rawLean.copy(
                state = if (
                    rawLean.state == RideDataState.VALID && !leanCalibrated
                ) {
                    RideDataState.DEGRADED
                } else {
                    rawLean.state
                },
            )
            val ble = when {
                isConnecting -> "CONNECTING" to RideDataState.DEGRADED
                connectionState == ConnectionState.CONNECTED ->
                    "BLE" to RideDataState.VALID
                connectionState == ConnectionState.SCANNING ->
                    "SCANNING" to RideDataState.DEGRADED
                else -> "BLE LOST" to RideDataState.UNAVAILABLE
            }
            val zone = TachometerZoneResolver.zone(rpm.displayValue, profile)
            return RideDashboardState(
                speed = speed,
                engineRpm = rpm,
                gear = RideGear.UNAVAILABLE,
                leanAngle = lean,
                engineTemperature = telemetry.vehicle.engineTemperature.toRideValue(),
                fuelLevel = telemetry.vehicle.fuelLevel.toRideValue(),
                ambientTemperature = telemetry.vehicle.ambientTemperature.toRideValue(),
                ambientLight = telemetry.ambientLight.toRideValue(),
                batteryVoltage = telemetry.batteryVoltage.toRideValue(),
                totalCurrent = telemetry.totalCurrent.toRideValue(),
                canState = telemetry.can1.state.toRideValue(),
                bleLabel = ble.first,
                bleState = ble.second,
                tachometerZone = zone,
                alerts = buildAlerts(
                    telemetry = telemetry,
                    rpm = rpm,
                    zone = zone,
                    connectionState = connectionState,
                    isConnecting = isConnecting,
                    profile = profile,
                ),
                profile = profile,
            )
        }

        private fun buildAlerts(
            telemetry: TelemetrySnapshot,
            rpm: RideValue<Double>,
            zone: TachometerZone,
            connectionState: ConnectionState,
            isConnecting: Boolean,
            profile: VehiclePerformanceProfile,
        ): List<RideAlert> {
            val alerts = telemetry.warnings
                .filter { it.active }
                .associateBy(
                    keySelector = DeviceWarning::code,
                    valueTransform = {
                        RideAlert(
                            id = it.code,
                            title = it.message,
                            severity = severity(it.severity),
                        )
                    },
                )
                .toMutableMap()
            if (zone == TachometerZone.RED && rpm.displayValue != null) {
                alerts.putIfAbsent(
                    "high_rpm",
                    RideAlert("high_rpm", "HIGH RPM", RideAlertSeverity.CRITICAL),
                )
            }
            val fuel = telemetry.vehicle.fuelLevel.toRideValue().displayValue
            val capacity = profile.fuelCapacityLiters
            val reserve = profile.fuelReserveLiters
            if (
                fuel != null &&
                capacity != null &&
                reserve != null &&
                capacity > 0 &&
                fuel <= reserve / capacity * 100
            ) {
                alerts.putIfAbsent(
                    "fuel_reserve",
                    RideAlert(
                        "fuel_reserve",
                        "FUEL RESERVE",
                        RideAlertSeverity.WARNING,
                    ),
                )
            }
            val ambient = telemetry.vehicle.ambientTemperature
                .toRideValue()
                .displayValue
            val iceThreshold = profile.iceWarningTemperatureCelsius
            if (ambient != null && iceThreshold != null && ambient < iceThreshold) {
                alerts.putIfAbsent(
                    "ice_warning",
                    RideAlert(
                        "ice_warning",
                        "ICE WARNING",
                        RideAlertSeverity.WARNING,
                    ),
                )
            }
            if (connectionState == ConnectionState.DISCONNECTED && !isConnecting) {
                alerts.putIfAbsent(
                    "ble_lost",
                    RideAlert(
                        "ble_lost",
                        "SVC CONNECTION LOST",
                        RideAlertSeverity.WARNING,
                    ),
                )
            }
            if (telemetry.can1.state.toRideValue().displayValue == null) {
                alerts.putIfAbsent(
                    "can_unavailable",
                    RideAlert(
                        "can_unavailable",
                        "CAN UNAVAILABLE",
                        RideAlertSeverity.WARNING,
                    ),
                )
            }
            if (telemetry.storage.sdCardState.toRideValue().displayValue == "missing") {
                alerts.putIfAbsent(
                    "sd_card_missing",
                    RideAlert(
                        "sd_card_missing",
                        "SD CARD MISSING",
                        RideAlertSeverity.WARNING,
                    ),
                )
            }
            return alerts.values.toList()
        }

        private fun severity(value: String): RideAlertSeverity = when (
            value.lowercase()
        ) {
            "critical" -> RideAlertSeverity.CRITICAL
            "warning" -> RideAlertSeverity.WARNING
            else -> RideAlertSeverity.INFO
        }
    }
}

enum class RideThemeMode {
    DAY,
    NIGHT,
    AUTOMATIC,
}

enum class RideResolvedTheme {
    DAY,
    NIGHT,
}

data class RideThemeThresholds(
    val nightEnterLux: Double,
    val dayEnterLux: Double,
) {
    companion object {
        val DEFAULT = RideThemeThresholds(
            nightEnterLux = 250.0,
            dayEnterLux = 650.0,
        )
    }
}

object RideThemeResolver {
    fun resolve(
        mode: RideThemeMode,
        ambientLight: RideValue<Double>,
        previous: RideResolvedTheme,
        thresholds: RideThemeThresholds = RideThemeThresholds.DEFAULT,
    ): RideResolvedTheme = when (mode) {
        RideThemeMode.DAY -> RideResolvedTheme.DAY
        RideThemeMode.NIGHT -> RideResolvedTheme.NIGHT
        RideThemeMode.AUTOMATIC -> {
            val lux = ambientLight.displayValue ?: return previous
            when (previous) {
                RideResolvedTheme.DAY ->
                    if (lux <= thresholds.nightEnterLux) {
                        RideResolvedTheme.NIGHT
                    } else {
                        RideResolvedTheme.DAY
                    }
                RideResolvedTheme.NIGHT ->
                    if (lux >= thresholds.dayEnterLux) {
                        RideResolvedTheme.DAY
                    } else {
                        RideResolvedTheme.NIGHT
                    }
            }
        }
    }
}

object RideMotionPolicy {
    fun durationMillis(
        reduceMotion: Boolean,
        standardDurationMillis: Int = 180,
    ): Int = if (reduceMotion) 0 else standardDurationMillis
}

File: software/mobile/ios/SVCMobile/Sources/Model/RideDashboard.swift
import Foundation

enum RideDataState: String, Equatable {
    case valid
    case stale
    case degraded
    case invalid
    case unavailable
}

struct RideValue<Value: Equatable>: Equatable {
    let value: Value?
    let unit: String
    let state: RideDataState

    var displayValue: Value? {
        switch state {
        case .valid, .degraded:
            value
        case .stale, .invalid, .unavailable:
            nil
        }
    }
}

extension Measurement where Value: Equatable {
    var rideValue: RideValue<Value> {
        let state: RideDataState
        if stale {
            state = .stale
        } else if !valid || value == nil {
            state = quality == .invalid ? .invalid : .unavailable
        } else {
            switch quality {
            case .good:
                state = .valid
            case .degraded:
                state = .degraded
            case .invalid:
                state = .invalid
            case .unavailable:
                state = .unavailable
            }
        }
        return RideValue(value: value, unit: unit, state: state)
    }
}

enum RideGear: String, CaseIterable, Equatable {
    case neutral = "N"
    case first = "1"
    case second = "2"
    case third = "3"
    case fourth = "4"
    case fifth = "5"
    case sixth = "6"
    case between = "BETWEEN"
    case unavailable

    var displayValue: String {
        switch self {
        case .unavailable:
            "—"
        default:
            rawValue
        }
    }
}

enum TachometerZone: String, Equatable {
    case normal
    case warning
    case red
    case unavailable
}

enum TachometerZoneResolver {
    static func zone(
        rpm: Double?,
        profile: VehiclePerformanceProfile
    ) -> TachometerZone {
        guard
            let rpm,
            profile.tachometerScaleMaxRpm != nil
        else {
            return .unavailable
        }
        if let red = profile.redZoneStartRpm, rpm >= Double(red) {
            return .red
        }
        if let warning = profile.warningStartRpm, rpm >= Double(warning) {
            return .warning
        }
        return .normal
    }

    static func zoneWithHysteresis(
        rpm: Double?,
        profile: VehiclePerformanceProfile,
        previous: TachometerZone,
        hysteresisRpm: Double = 80
    ) -> TachometerZone {
        guard let rpm else { return .unavailable }
        let raw = zone(rpm: rpm, profile: profile)
        switch previous {
        case .red:
            if
                let red = profile.redZoneStartRpm,
                rpm >= Double(red) - hysteresisRpm
            {
                return .red
            }
        case .warning:
            if
                let red = profile.redZoneStartRpm,
                rpm >= Double(red)
            {
                return .red
            }
            if
                let warning = profile.warningStartRpm,
                rpm >= Double(warning) - hysteresisRpm
            {
                return .warning
            }
        case .normal, .unavailable:
            break
        }
        return raw
    }
}

struct LeanExtrema: Equatable {
    private(set) var maximumLeftDegrees: Double = 0
    private(set) var maximumRightDegrees: Double = 0

    mutating func observe(_ leanDegrees: Double?) {
        guard let leanDegrees else { return }
        if leanDegrees < 0 {
            maximumLeftDegrees = max(maximumLeftDegrees, abs(leanDegrees))
        } else {
            maximumRightDegrees = max(maximumRightDegrees, leanDegrees)
        }
    }

    mutating func resetIfStationary(speed: RideValue<Double>) -> Bool {
        guard speed.displayValue == 0 else { return false }
        maximumLeftDegrees = 0
        maximumRightDegrees = 0
        return true
    }
}

struct MotorcycleMountingTransform: Codable, Equatable {
    let rollDegrees: Double
    let pitchDegrees: Double
    let yawDegrees: Double
    let zeroOffsetDegrees: Double
}

enum MotorcycleCalibrationPolicy {
    static func canCalibrate(
        speed: RideValue<Double>,
        demoMode: Bool
    ) -> Bool {
        demoMode && speed.displayValue == 0
    }
}

enum RideAlertSeverity: Int, Comparable {
    case info = 0
    case warning = 1
    case critical = 2

    static func < (lhs: RideAlertSeverity, rhs: RideAlertSeverity) -> Bool {
        lhs.rawValue < rhs.rawValue
    }
}

struct RideAlert: Equatable, Identifiable {
    let id: String
    let title: String
    let severity: RideAlertSeverity
}

struct RideDashboardState: Equatable {
    let speed: RideValue<Double>
    let engineRpm: RideValue<Double>
    let gear: RideGear
    let leanAngle: RideValue<Double>
    let engineTemperature: RideValue<Double>
    let fuelLevel: RideValue<Double>
    let ambientTemperature: RideValue<Double>
    let ambientLight: RideValue<Double>
    let batteryVoltage: RideValue<Double>
    let totalCurrent: RideValue<Double>
    let canState: RideValue<String>
    let bleLabel: String
    let bleState: RideDataState
    let tachometerZone: TachometerZone
    let alerts: [RideAlert]
    let profile: VehiclePerformanceProfile

    var activeAlert: RideAlert? {
        alerts.max {
            if $0.severity == $1.severity {
                return $0.id > $1.id
            }
            return $0.severity < $1.severity
        }
    }

    static func build(
        telemetry: TelemetrySnapshot,
        connectionState: ConnectionState,
        isConnecting: Bool,
        profile: VehiclePerformanceProfile,
        leanCalibrated: Bool = false
    ) -> RideDashboardState {
        let speed = telemetry.vehicle.speed.rideValue
        let rpm = telemetry.vehicle.engineRpm.rideValue
        let rawLean = telemetry.leanAngle.rideValue
        let lean = RideValue(
            value: rawLean.value,
            unit: rawLean.unit,
            state: rawLean.state == .valid && !leanCalibrated
                ? .degraded
                : rawLean.state
        )
        let ble: (String, RideDataState) = if isConnecting {
            ("CONNECTING", .degraded)
        } else {
            switch connectionState {
            case .connected:
                ("BLE", .valid)
            case .scanning:
                ("SCANNING", .degraded)
            case .disconnected:
                ("BLE LOST", .unavailable)
            }
        }
        let zone = TachometerZoneResolver.zone(
            rpm: rpm.displayValue,
            profile: profile
        )
        let alerts = buildAlerts(
            telemetry: telemetry,
            rpm: rpm,
            zone: zone,
            connectionState: connectionState,
            isConnecting: isConnecting,
            profile: profile
        )
        return RideDashboardState(
            speed: speed,
            engineRpm: rpm,
            gear: .unavailable,
            leanAngle: lean,
            engineTemperature: telemetry.vehicle.engineTemperature.rideValue,
            fuelLevel: telemetry.vehicle.fuelLevel.rideValue,
            ambientTemperature: telemetry.vehicle.ambientTemperature.rideValue,
            ambientLight: telemetry.ambientLight.rideValue,
            batteryVoltage: telemetry.batteryVoltage.rideValue,
            totalCurrent: telemetry.totalCurrent.rideValue,
            canState: telemetry.can1.state.rideValue,
            bleLabel: ble.0,
            bleState: ble.1,
            tachometerZone: zone,
            alerts: alerts,
            profile: profile
        )
    }

    private static func buildAlerts(
        telemetry: TelemetrySnapshot,
        rpm: RideValue<Double>,
        zone: TachometerZone,
        connectionState: ConnectionState,
        isConnecting: Bool,
        profile: VehiclePerformanceProfile
    ) -> [RideAlert] {
        var alerts = telemetry.warnings
            .filter(\.active)
            .map {
                RideAlert(
                    id: $0.code,
                    title: $0.message,
                    severity: severity($0.severity)
                )
            }
        var codes = Set(alerts.map(\.id))

        if zone == .red, rpm.displayValue != nil {
            append(
                RideAlert(id: "high_rpm", title: "HIGH RPM", severity: .critical),
                to: &alerts,
                codes: &codes
            )
        }
        if
            let fuel = telemetry.vehicle.fuelLevel.rideValue.displayValue,
            let capacity = profile.fuelCapacityLiters,
            let reserve = profile.fuelReserveLiters,
            capacity > 0,
            fuel <= reserve / capacity * 100
        {
            append(
                RideAlert(
                    id: "fuel_reserve",
                    title: "FUEL RESERVE",
                    severity: .warning
                ),
                to: &alerts,
                codes: &codes
            )
        }
        if
            let ambient = telemetry.vehicle.ambientTemperature.rideValue.displayValue,
            let threshold = profile.iceWarningTemperatureCelsius,
            ambient < threshold
        {
            append(
                RideAlert(id: "ice_warning", title: "ICE WARNING", severity: .warning),
                to: &alerts,
                codes: &codes
            )
        }
        if connectionState == .disconnected, !isConnecting {
            append(
                RideAlert(id: "ble_lost", title: "SVC CONNECTION LOST", severity: .warning),
                to: &alerts,
                codes: &codes
            )
        }
        if telemetry.can1.state.rideValue.displayValue == nil {
            append(
                RideAlert(id: "can_unavailable", title: "CAN UNAVAILABLE", severity: .warning),
                to: &alerts,
                codes: &codes
            )
        }
        if telemetry.storage.sdCardState.rideValue.displayValue == "missing" {
            append(
                RideAlert(id: "sd_card_missing", title: "SD CARD MISSING", severity: .warning),
                to: &alerts,
                codes: &codes
            )
        }
        return alerts
    }

    private static func append(
        _ alert: RideAlert,
        to alerts: inout [RideAlert],
        codes: inout Set<String>
    ) {
        guard codes.insert(alert.id).inserted else { return }
        alerts.append(alert)
    }

    private static func severity(_ value: String) -> RideAlertSeverity {
        switch value.lowercased() {
        case "critical":
            .critical
        case "warning":
            .warning
        default:
            .info
        }
    }
}
